### PR TITLE
Don't allocate 4kB and Stream.CopyTo for empty bodies

### PR DESCRIPTION
**Pull request by benaadams**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/NancyFx/Nancy/pulls) open
- [x] I have verified that I am following the Nancy [code style guidelines](https://github.com/NancyFx/Nancy/blob/45238076ad0b7f6ecabd6bae8469e30458d02efe/CONTRIBUTING.md#style-guidelines)
- [ ] I have provided test coverage for my change (where applicable)

### Description

Stream.CopyTo on Full Framework and Mono will allocate a buffer of the `RequestStream.BufferSize` for the Stream.CopyTo which is 4kB (CoreClr will use ArrayPool for this); however this can be entirely skipped when `ContentLength == 0` or there is no `Content-Length` or `Transfer-Encoding` (e.g. `chunked`)

This means for example "GET" requests with no body will no longer allocate 4kB by default.

### CODE REVIEW

**Comment by benaadams** [comment]
This change and using a Release build https://github.com/NancyFx/Nancy/issues/2897 increase the Mono throughput for Nancy plaintext on Techempower by x 4.5

File: src/Nancy/Owin/NancyMiddleware.cs [hunk 1]
```diff
@@ -80,7 +80,9 @@ public static MidFunc UseNancy(NancyOptions options = null)
 
                         var url = CreateUrl(owinRequestHost, owinRequestScheme, owinRequestPathBase, owinRequestPath, owinRequestQueryString);
 
-                        var nancyRequestStream = new RequestStream(owinRequestBody, ExpectedLength(owinRequestHeaders), StaticConfiguration.DisableRequestStreamSwitching ?? false);
+                        var expectedLength = ExpectedLength(owinRequestHeaders);
+                        // If length is 0 just use empty memory stream; as there is no body
+                        var nancyRequestStream = (expectedLength == 0) ? (Stream)new MemoryStream() : new RequestStream(owinRequestBody, ExpectedLength(owinRequestHeaders) ?? 0, StaticConfiguration.DisableRequestStreamSwitching ?? false);
```
Comment (khellang) [inline src/Nancy/Owin/NancyMiddleware.cs]: We should probably use `expectedLength` instead of calling `ExpectedLength(owinRequestHeaders)` again?

Comment (benaadams) [reply src/Nancy/Owin/NancyMiddleware.cs]: Lol; yes

Comment (benaadams) [reply src/Nancy/Owin/NancyMiddleware.cs]: Changed

**Comment by khellang** [comment]
Thanks! ❤️

**Pull request merged**